realtek: add support for NEXT 7255GH switch

koreapyj opened the pull request:
NEXT 7255GH is unmanaged 5 port 2.5GbE switch.

Hardware:
- SoC: Realtek RTL9303 SoC
- RAM: DDR3 128MB
- Flash: Macronix MX25L12835F (SPI-NOR 16MB)
- Ethernet: RTL8226 5x 2.5GbE PHY
- UART: [GND, RX, TX, 3.3V] (115200)

Known issues:
- Packet loss on MTU over 1472
- DHCP client not working
- No MAC address
- No recovery methods except UART console
- Switch malfunction. Device network works but switching is not working for unknown reason.

Note:
- Rev. A2 only. Rev.B has RTL8371 and no memory.

oliv3r commented on target/linux/realtek/files-5.10/drivers/net/dsa/rtl83xx/common.c:
```diff
@@ -316,6 +316,7 @@ static int __init rtl83xx_mdio_probe(struct rtl838x_switch_priv *priv)
 	for_each_node_by_name(dn, "port") {
 		phy_interface_t interface;
 		u32 led_set;
+		u32 led_num;
```
This probably should be its own commit, though with the lack of a proper led driver atm I see what you are doing. Would this be working in any way with https://gitlab.com/olliver/openwrt/openwrt/-/commit/7df2e63f7bf2a1b7de18007e9cb209ea92e9c415 ? I know that that patch is a hack too; but both these are just while we wait for sander's led driver stuff ...

oliv3r commented:
@koreapyj I've added support to my WIP tree under commit https://gitlab.com/olliver/openwrt/openwrt/-/commit/849c76aef7cd2529d3b39e8ae20f19be90c659e2 which I'm sure will change a little before its ready for merge.

Can you verify that things check out, as we don't have any information of this switch over on the wiki. What I'm wondering about is if the power led is always on, or is actually the sys_led, and if there's a reset switch.

The problems you've listed seem to be generic rtl93xx based problems, and is being investigated/discussed on the forum in the [realtek thread](https://forum.openwrt.org/t/support-for-rtl838x-based-managed-switches/57875) so hopefully this will be resolved. Btw, dhcp works fine, but only if it is requested before networking starts up. So either hardcoded in the network config script as default (02-network) or setting it before the network script runs (when booting initramfs). Booting from flash and committing the setting should also work.

The led patches are not needed anymore, though the MR still hasn't been merged sadly :(

The Ethernet driver fix looks interesting and I'll investigate it and will put it in my tree as well. It seems very logical by all means.

koreapyj committed:
realtek: add support for NEXT 7255GH switch

- DHCP (client) not working
- switch does not switching
- something weird happening on somewhere

koreapyj commented:
Rebased on current 'main'.

- Now MTU related issues are gone.
- But DHCP is still not working.
- Switch is still not switching.
- And mdio-bus fail with error -5
    ```
    [    0.904833] RTL8226B-CG_RTL8221B-CG 2.5Gbps PHY: probe of mdio-bus:01 failed with error -5
    [    1.005743] RTL8226B-CG_RTL8221B-CG 2.5Gbps PHY: probe of mdio-bus:02 failed with error -5
    [    1.106653] RTL8226B-CG_RTL8221B-CG 2.5Gbps PHY: probe of mdio-bus:03 failed with error -5
    [    1.207564] RTL8226B-CG_RTL8221B-CG 2.5Gbps PHY: probe of mdio-bus:04 failed with error -5
    [    1.329525] RTL8226B-CG_RTL8221B-CG 2.5Gbps PHY: probe of mdio-bus:05 failed with error -5
    ```

It's a little off topic but @oliv3r 's branch also not working but symptoms are different.

- mdio-bus fail with error -5 (same)
- Kernel does not detect link status, always show LOWERLAYERDOWN.

koreapyj commented:
This device is confirmed to identical with Trendnet TEG-S350 and Zyxel MG-105. PCB layout, case is identical.
NEXT-7258GH is 8 port version and also identical with TEG-S380 and MG-108.
